Add a document describing expectations for pull requests

larsks opened the pull request:
The team recently expressed interest in a document that sets forth our expectations for pull requests. This is my attempt at that document.

I've tried to avoid just rewriting the hundreds of documents on this topic that are already out there. I've provided several links to documents that I think provide more detailed coverage of the same ideas in this document.

larsks committed:
Add a document describing expectations for pull requests

The team recently expressed interest in a document that sets forth our
expectations for pull requests. This is my attempt at that document.

naved001 commented on specs/0007-pull-requests.md:
```diff
@@ -0,0 +1,144 @@
+  If people are taking the time to review your code, you should
+  respond promptly to those reviews. This may mean you need to pivot
+  away from working on an exciting new feature in order to address the
+  requested changes.
```
Can we add a bit about not force pushing commits during the review phase of a PR? and maybe squash the commits when it's ready to be merged.

rob-baron replied:
we also should consider a list of useful git command.

naved001 replied:
> useful git command.

I dont know, this doesn't seem like the right place to document it.

larsks replied:
Yeah, this isn't meant to be a "how to use git" document. There are already lots of those out there.

larsks replied:
I've added a note to the document about appending new commits instead of rebasing.

naved001 reviewed (commented):
While it looks good to me, we should definitely wait on @rob-baron's review of this.

rob-baron commented on specs/0007-pull-requests.md:
```diff
@@ -0,0 +1,144 @@
+a `pre-commit` hook. Your pull request should pass any unit tests that
+exist in the repository (and if you are introducing a new feature,
+your pull request should include tests appropriate to that feature).
+
```
Somewhere we need to mention:
1) More specific lint requirements
2) Which specific pretty printer (ie 'black') should be used and its specific configuration if it is not the default
3) Potentially a project that has just a skeleton that gives the basic support for the tools that they are expected to use.

Pretty printing is useful as it enables GitHub to make better diffs which highlights the changes that were made, as opposed to formatting changes.

larsks replied:
> Somewhere we need to mention:

I don't think these are appropriate to this document, because they're all going to vary by project. We have things written in Python, Go, Shell, YAML, JSON, Markdown, etc, and each language is going to have a different set of tools.

Additionally, it's not my goal to be prescriptive here: different projects may have different requirements (for example, a project that lives within the OpenStack ecosystem may use a specific set of tooling that's not necessary or desirable in another project).

I think we want people to have flexibility in their tooling as long as they meet the requirement of submitting changes that are free of simple errors.

larsks replied:
I've added a node to the document about the utility of code formatting tools.

rob-baron commented on specs/0007-pull-requests.md:
```diff
@@ -0,0 +1,144 @@
+  Don't just tell people that something is broken. Explain why you
+  believe something needs changing, and consider providing examples
+  where appropriate.
+
```
Constructive feedback can also extend to the structure of the PR.

larsks replied:
I agree. Do you think that requires a change to the document?

rob-baron replied:
Upon further reflection, I think this needs to flesh out some of the general possibilities.  Like, problems with the specification/user cases, problems with the design, problems with the implementation.  It could be problems with how the code is organized or even how the PR is organized just to name a few.

Personally, I am structuring my PRs/work so that it will be as easy as possible to get them through the review process.  It's hard to describe my thought structure yet, but it is getting there.

rob-baron replied:
I'm ok with this for now.

larsks committed:
Add a note about the utility of code formatting tools

@rob-baron points out correctly that enforcing a code formatting
standard (e.g., using black or go fmt, etc) makes diffs more
meaningful..

larsks committed:
Add note about appending rather than rebasing

Appending commits to a pull request rather than rebasing it makes it
easier for reviews to see what has changed.

larsks committed:
Add some examples of commit messages to avoid

larsks committed:
Note that syntax and linting errors aren't a good use of reviewer time

rob-baron commented:
we should follow up with this listing commonly used tooling for different languages (python, go, c, Perl, ... )
